ml-agents/mlagents/trainers/settings.py

```diff
@@ -0,0 +1,365 @@
+import attr
+import cattr
+from typing import Dict, Optional, List, Any, DefaultDict, Mapping
+from enum import Enum
+import collections
+import argparse
+
+from mlagents.trainers.cli_utils import StoreConfigFile, DetectDefault, parser
+from mlagents.trainers.cli_utils import load_config
+from mlagents.trainers.exception import TrainerConfigError
+from mlagents.trainers.models import LearningRateSchedule, EncoderType
+
+
+def check_and_structure(key: str, value: Any, class_type: type) -> Any:
+    attr_fields_dict = attr.fields_dict(class_type)
+    if key not in attr_fields_dict:
+        raise TrainerConfigError(
+            f"The option {key} was specified in your YAML file for {class_type.__name__}, but is invalid."
+        )
+    # Apply cattr structure to the values
+    return cattr.structure(value, attr_fields_dict[key].type)
+
+
+def strict_to_cls(d: Mapping, t: type) -> Any:
+    if d is None:
+        return None
+    d_copy: Dict[str, Any] = {}
+    d_copy.update(d)
+    for key, val in d_copy.items():
+        d_copy[key] = check_and_structure(key, val, t)
+    return t(**d_copy)
+
+
+def trainer_settings_to_cls(d: Mapping, t: type) -> Any:
+    if d is None:
+        return None
+    d_copy: Dict[str, Any] = {}
+    d_copy.update(d)
+
+    for key, val in d_copy.items():
+        if attr.has(type(val)):
+            # Don't convert already-converted attrs classes.
+            continue
+        if key == "hyperparameters":
+            if "trainer_type" not in d_copy:
+                raise TrainerConfigError(
+                    "Hyperparameters were specified but no trainer_type was given."
+                )
+
+            else:
+                d_copy[key] = strict_to_cls(
+                    d_copy[key], TrainerSettings.to_settings(d_copy["trainer_type"])
+                )
+        elif key == "reward_signals":
+            d_copy[key] = rewardsignal_settings_to_cls(val)
+        elif key == "max_steps":
+            d_copy[key] = int(
+                float(val)
+            )  # In some configs, max steps was specified as a float
+        else:
+            d_copy[key] = check_and_structure(key, val, t)
+    return t(**d_copy)
+
+
+def rewardsignal_settings_to_cls(d: Mapping) -> Any:
+    if d is None:
```

Comment:
Should / could we just prevent `None` from being passed here?  If not, `d` should be an `Optional[Mapping]`.

Reply:
True, these should never be None unless we have a malformed config (e.g. if user puts `env_settings: null` in their config). I now throw an error if a None (or non-Mapping) is received.